### PR TITLE
328 final

### PR DESCRIPTION
**Pull request by withrocks**
A large refactoring where we use the name Work instead of Task. Also includes fixes for enabling assigning samples to a workflow and get into a workbatch view.

### CODE REVIEW

File: src/clims/api/serializers/models/work_batch_details.py
Comment (withrocks) [inline src/clims/api/serializers/models/work_batch_details.py:15]: All of this stuff I'll revisit next. It was convenient for me to keep it in this PR.

File: src/clims/migrations/0022_workbatch_input_containers.py
Comment (withrocks) [inline src/clims/migrations/0022_workbatch_input_containers.py:14]: This change wasn't directly changed to this PR, but I'm following up on this in the next PR, when I focus on the work batch itself.

File: src/clims/services/workflow/camunda/client.py
Comment (withrocks) [inline src/clims/services/workflow/camunda/client.py]: I'll fix this.

File: src/sentry/static/sentry/app/redux/reducers/workBatchDetails.js
Comment (withrocks) [inline src/sentry/static/sentry/app/redux/reducers/workBatchDetails.js:15]: This I'll have to revisit in the next PR. I think the component that requires this should do this. But currently this breaks the  transition component (CLIMS-492). But since it was working with mocked data anyways (and I'm the one that's going to fix it) I felt that it was OK.

File: src/sentry/static/sentry/app/views/availableWorkDetails/index.jsx
Comment (withrocks) [inline src/sentry/static/sentry/app/views/availableWorkDetails/index.jsx]: I'll fix these comments.

File: src/clims/migrations/0022_workbatch_input_containers.py
Comment (GitEdvard) [inline src/clims/migrations/0022_workbatch_input_containers.py:19]: I guess that container is for the workbatches where work-unit is set to container? For ordinary workbatches, handling samples, input container would be a derived property from the list of substances. I think it might be confusing to both use derived input containers (in ordinary workbatches) and input container meant for workbatches handling containers only.

**Comment by GitEdvard** [comment]
I think that something is wrong here. The last 5 commits seem to be empty.

**Comment by GitEdvard** [comment]
Never mind, now they are back!

File: src/clims/api/endpoints/work_definition_details.py
Comment (GitEdvard) [inline src/clims/api/endpoints/work_definition_details.py:12]: There already exist a class for this that is named work_batch_details_definition.py

**Comment by withrocks** [comment]
I think it might have had to do with rebases.

File: src/clims/models/workunit.py
Comment (GitEdvard) [inline src/clims/models/workunit.py:108]: This sound for me like a work unit that has not yet been saved. Isn't work-unit sufficient for this?

File: src/clims/services/workflow/camunda/client.py
Comment (GitEdvard) [inline src/clims/services/workflow/camunda/client.py:103]: It should be a serializer doing this

File: src/clims/api/endpoints/work_batch.py
Comment (GitEdvard) [inline src/clims/api/endpoints/work_batch.py:41]: I'm wondering if it in any cases would be desirable to just create a workbatch and not doing anything else?

File: src/clims/api/endpoints/work_definition_details.py
Comment (GitEdvard) [inline src/clims/api/endpoints/work_definition_details.py:24]: It's different serializers for work definition and work definiton details

Comment (GitEdvard) [reply src/clims/api/endpoints/work_definition_details.py:24]: I find myself much more comfortable to use the word "step" instead of work definition, just because it's shorter. If we want to be really consistent, it should be "WorkBatchDefinitionSerializer" and "WorkBatchDetailsDefinitionSerializer". Should Details come before Definition? I rather quickly find this very awkward to write, pronounce and reason about.

File: src/clims/api/endpoints/work_units.py
Comment (GitEdvard) [inline src/clims/api/endpoints/work_units.py:16]: Endpoint classes are so far written in singular!

File: src/clims/api/serializers/models/workunit.py
Comment (GitEdvard) [inline src/clims/api/serializers/models/workunit.py:1]: There is an inconsistency of the file names, wheather it is an underscore between work_unit or not. I prefer not to have underscore, but we should first of all have consistency here.

File: tests/clims/api/endpoints/test_work_units.py
Comment (GitEdvard) [inline tests/clims/api/endpoints/test_work_units.py:37]: The last thing, "DataEntry", is a mystery for me. This is no work definition by the way! It's a workflow.

Comment (GitEdvard) [reply tests/clims/api/endpoints/test_work_units.py:37]: In other parts of the code, I've refered work defintion by its full class name

File: src/clims/models/workunit.py
Comment (GitEdvard) [inline src/clims/models/workunit.py:71]: As I see it, a new work unit instance is created in the clims db for each external work unit (?). In that case, the inherent id of work unit should suffice to differentiate between work units.

Comment (GitEdvard) [reply src/clims/models/workunit.py:71]: I'm afraid that there is a risk of fetching of e.g. substance from a workbatch is going to be cumbersome. We have workbatch --> get work units --> get substance. In between the client developer has to dig deep to understand external_work_unit_id, and external_workflow_instance_id. (why use these instead of inherent ids?) My feeling is that this has to be encapsulated in some way.

File: src/sentry/static/sentry/app/components/workUnit/processListItem.jsx
Comment (GitEdvard) [inline src/sentry/static/sentry/app/components/workUnit/processListItem.jsx:3]: I cannot interpret what processListItem does based on its name. Process is unclear to me, is it equivalent to a workflow, or is it bound to a workunit and a workflow? process list item doesn't make sense.

File: src/sentry/static/sentry/app/redux/actions/workDefinition.js
Comment (GitEdvard) [inline src/sentry/static/sentry/app/redux/actions/workDefinition.js:8]: As general rule, I think it's best not to add things before they are used anywhere. Going back here in 3 months, it's very hard to know if the code should be saved or is outdated.

File: src/sentry/static/sentry/app/redux/reducers/availableWorkUnit.js
Comment (GitEdvard) [inline src/sentry/static/sentry/app/redux/reducers/availableWorkUnit.js:1]: The name available work unit doesn't make sense. Available work definitions (steps) are the ones with samples queuing to it. As it goes for Clarity, all samples are available to everyone.

File: src/sentry/static/sentry/app/views/availableWorkDetails/index.jsx
Comment (GitEdvard) [inline src/sentry/static/sentry/app/views/availableWorkDetails/index.jsx:1]: This must be the view showing when user has clicked on a single workstep to see what samples in it?

**Review by GitEdvard — APPROVED** [review]
Nice work! I have some comments, although I think it's more important to get this into the code base, since it's blocking other work.

File: src/clims/migrations/0022_workbatch_input_containers.py
Comment (withrocks) [reply src/clims/migrations/0022_workbatch_input_containers.py:19]: The idea here was that this is the list of containers that need to be shown on the left in the transition component. I think that i was not thinking clearly when I added this! I'd prefer that we just remove it again when we tackle CLIMS-482 (in interest of time).

File: src/clims/api/endpoints/work_definition_details.py
Comment (withrocks) [reply src/clims/api/endpoints/work_definition_details.py:12]: I see :+1:. I threw in a cleanup ticket under the CLIMS-483 epic.

File: src/clims/models/workunit.py
Comment (withrocks) [reply src/clims/models/workunit.py:108]: Yeah, or it's one that has been exists in the workflow engine but not in our database.

I used only work unit at first, but since it's not just a regular unsaved Django model, that made it confusing to reason about. I had to have a special field on the model with info that only makes sense when it's a workunit that hasn't been saved yet.

I think actually a great way to resolve both this and the general save issue that you have been talking about (where we are requried to save the django objects in the db in order to save it's relations), is to have higher level business classes for everything, not only substances. They could then take care of all kinds of stuff similar to this. I think we should discuss that in detail when we've implemented the workbatch view. I've added it to CLIMS-505 so we don't forget it.

File: src/clims/services/workflow/camunda/client.py
Comment (withrocks) [reply src/clims/services/workflow/camunda/client.py:103]: That's a good point. I hadn't thought of it because I'm interacting with another system.

File: src/clims/api/endpoints/work_batch.py
Comment (withrocks) [reply src/clims/api/endpoints/work_batch.py:41]: Right, in this case I just needed a service-style call, one that takes care of a whole business action for me. But it might be needed later.

File: src/clims/api/endpoints/work_definition_details.py
Comment (withrocks) [reply src/clims/api/endpoints/work_definition_details.py:24]: I agree that the naming is not great. I've added an epic for dealing with this at a later point, but we should fix the things that are confusing as soon as we can (although I think we need to merge this one without further change, unless there's something really wrong :)).

I'm also still wondering if we shouldn't call this something else, e.g. Step or UserTask instead. The problem I see with Step, is that we also have non-usertask steps in workflows, so it may be confusing too.

But if this was renamed either, I think we would need:
* [UserTask/Step]DefinitionSerializer (the definition of how the user task should behave, serialized in a list)
* [UserTask/Step]DefinitionDetailsSerializer (the same, serialized in a details view)

Details I think should always be after the Definition word, because it's used to mean that this is a details entry, but not a list entry.

I think we should take this up when we tackle the next task related to this!

File: src/clims/api/endpoints/work_units.py
Comment (withrocks) [reply src/clims/api/endpoints/work_units.py:16]: Right :| I can fix this in a follow-up PR.

File: src/clims/api/serializers/models/workunit.py
Comment (withrocks) [reply src/clims/api/serializers/models/workunit.py:1]: I agree.

File: tests/clims/api/endpoints/test_work_units.py
Comment (withrocks) [reply tests/clims/api/endpoints/test_work_units.py:37]: The first part refers to the Workflow class. That's indeed just a regular class. Then that workflow has a "UserTask" in Camunda. It does not necessarily relate to any Python class. Maybe it should, but that would mean that the user needs to write more code. I would prefer that they would only have to write a BPMN diagram and minimum Python code to get going.

Comment (withrocks) [reply tests/clims/api/endpoints/test_work_units.py:37]: ... I looked a bit more into this. I think it's how it used to work before and needs to be further fixed. The test doesn't even seem to run here (although everything ran on github). I'll take a peek.

Comment (withrocks) [reply tests/clims/api/endpoints/test_work_units.py:37]: OK, I think I'll just fix this in another PR (CLIMS-507, assigned to me). This is how it was coupled up to begin with but it should be the full class name of the work definition, but it is the work definition

File: src/clims/models/workunit.py
Comment (withrocks) [reply src/clims/models/workunit.py:71]: This is the key that's required when interfacing with the external workflow engine. The plugin dev will not have to worry about this - but we will :) Sounds good to make it easier to deal with.

File: src/sentry/static/sentry/app/components/workUnit/processListItem.jsx
Comment (withrocks) [reply src/sentry/static/sentry/app/components/workUnit/processListItem.jsx:3]: This is an old name that I forgot to change. I'll throw it in the tech debt pile.

File: src/sentry/static/sentry/app/redux/actions/workDefinition.js
Comment (withrocks) [reply src/sentry/static/sentry/app/redux/actions/workDefinition.js:8]: Agreed. The actions is used. It's just the list endpoint that's not used. It's a hack though, and I was confused by this myself. Should just be null.

File: src/sentry/static/sentry/app/redux/reducers/availableWorkUnit.js
Comment (withrocks) [reply src/sentry/static/sentry/app/redux/reducers/availableWorkUnit.js:1]: The samples may be queued elsewhere, but this the list of work units that are ready for work. But the name is not clear, I agree.

File: src/sentry/static/sentry/app/views/availableWorkDetails/index.jsx
Comment (withrocks) [reply src/sentry/static/sentry/app/views/availableWorkDetails/index.jsx:1]: Right. Technically, it will list work units that are in it, but the user sees samples.

**Comment by withrocks** [comment]
Thanks for reading all of this again so thoroughly!

**Pull request merged**
